### PR TITLE
use ioutil.TempDir instead of a static temporary directory in kubelet…

### PR DESCRIPTION
**Pull request by jijun2**
… cni-test.go
related to #15176

### CODE REVIEW

File: pkg/kubelet/network/cni/cni_test.go [hunk 1]
```diff
@@ -42,10 +42,17 @@ import (
...
-const testNetworkConfigPath = "/tmp/fake/plugins/net/cni"
+func tmpDirOrDie() string {
+	dir, err := ioutil.TempDir(os.TempDir(), "kubelet-test")
+	if err != nil {
+		panic(fmt.Sprintf("error creating tmp dir: %v", err))
+	}
+	return path.Join(dir, "plugins", "net", "cni")
+}
+
 const testVendorCNIDirPrefix = "/tmp"
```
Comment (ixdy) [inline pkg/kubelet/network/cni/cni_test.go]: probably should handle this one too.

File: pkg/kubelet/network/cni/cni_test.go [hunk 2]
```diff
@@ -42,10 +42,17 @@ import (
 )
 
 // The temp dir where test plugins will be stored.
-const testNetworkConfigPath = "/tmp/fake/plugins/net/cni"
+func tmpDirOrDie() string {
+	dir, err := ioutil.TempDir(os.TempDir(), "kubelet-test")
```
Comment (ixdy) [inline pkg/kubelet/network/cni/cni_test.go]: minor nit, but "cni-test" might be a better name than "kubelet-test"

**Comment by ixdy** [comment]
also FYI the Shippable failure looks like a flake unrelated to this change.

**Comment by jijun2** [comment]
@ixdy, I have changed.

File: pkg/kubelet/network/cni/cni_test.go [hunk 3]
```diff
@@ -103,13 +108,12 @@ echo -n "{ \"ip4\": { \"ip\": \"10.1.0.23/24\" } }"
 	f.Close()
 }
 
-func tearDownPlugin(plugName string, vendorName string) {
+func tearDownPlugin(testNetworkConfigPath string, testVendorCNIDirPrefix string) {
```
Comment (ixdy) [inline pkg/kubelet/network/cni/cni_test.go]: this isn't deleting testNetworkConfigPath

File: pkg/kubelet/network/cni/cni_test.go [hunk 4]
```diff
@@ -42,10 +42,15 @@ import (
...
 
 // The temp dir where test plugins will be stored.
-const testNetworkConfigPath = "/tmp/fake/plugins/net/cni"
-const testVendorCNIDirPrefix = "/tmp"
+func tmpDirOrDie() string {
+	dir, err := ioutil.TempDir(os.TempDir(), "cni-test")
+	if err != nil {
+		panic(fmt.Sprintf("error creating tmp dir: %v", err))
+	}
+	return path.Join(dir, "plugins", "net", "cni")
```
Comment (ixdy) [inline pkg/kubelet/network/cni/cni_test.go]: might be better to have this just return `dir`, then call this function only once in TestCNIPlugin, using it to construct both other directories.

then the cleanup only needs to worry about this one root directory, too.

**Comment by jijun2** [comment]
@ixdy ,Thank you for your comments, I will change.

**Comment by ashcrow** [comment]
@jijun2 This is looking good!

### Nit
The commit message doesn't reflect the change as well as it could. It's referencing one specific change rather than noting the fix of the issue. Maybe something like "test: using ioutil.TempDir in unit tests per #15176" or something like that may be a more representative message.


### Future Idea (change not required)
There also seems to be a lot of the same kind of code over and over. I think this is OK but in the future we may wan to look at providing a reusable temporary directory creator that can be used in tests:

```go
// makeTmpDir creates a temporary directory based on the prefix given. It returns the temporary
// directory path along with a deferable function to remove the directory. If an error occurs during
// creation t.Fatal is called and execution is halted.
func makeTmpDir(t *testing.T, prefix string) (string, func()) {
    tmpDir, err := ioutil.TempDir(os.TempDir(), prefix)
    if err != nil {
        t.Fatal("can't make a temp dir: %v", err)
    }
    return tmpDir, func() { defer os.RemoveAll(tmpDir) }
}
```

Example:
```go
tmpDir, cleanUp := makeTmpDir(t, "test")
defer cleanUp()
/...
```

**Comment by ashcrow** [comment]
@k8s-bot test this please

**Comment by jijun2** [comment]
@ashcrow ,Sorry ,the commit message isn't changed, it is only squash the commits.

**Comment by ashcrow** [comment]
@jijun2 I understand, no need to apologize :smile: If you can use amend to update the commit message that would do it. You'll have to force push to the repository afterwards.

**Comment by jijun2** [comment]
@ashcrow, OK！

File: pkg/volume/aws_ebs/aws_ebs_test.go [hunk 5]
```diff
@@ -118,9 +135,9 @@ func TestPlugin(t *testing.T) {
 	if builder == nil {
 		t.Errorf("Got a nil Builder")
 	}
-
+	volPath := path.Join(tmpDir, "pods/poduid/volumes/kubernetes.io~aws-ebs/vol1")
```
Comment (ashcrow) [inline pkg/volume/aws_ebs/aws_ebs_test.go]: Doesn't require a fix, just FYI:

On https://github.com/jijun2/kubernetes/blob/fix-cni_test/pkg/kubelet/network/cni/cni_test.go#L172 you split up into chunks. Just be aware you are sometimes doing that and sometimes joining a path with a forward slash'd path.

File: pkg/volume/cephfs/cephfs_test.go [hunk 6]
```diff
@@ -71,8 +83,9 @@ func TestPlugin(t *testing.T) {
 	if builder == nil {
 		t.Errorf("Got a nil Builder: %v")
 	}
+	volpath := path.Join(tmpDir, "pods/poduid/volumes/kubernetes.io~cephfs/vol1")
 	path := builder.GetPath()
-	if path != "/tmp/fake/pods/poduid/volumes/kubernetes.io~cephfs/vol1" {
+	if path != volpath {
```
Comment (ashcrow) [inline pkg/volume/cephfs/cephfs_test.go]: :+1

**Comment by ashcrow** [comment]
This looks good to me.

File: pkg/volume/aws_ebs/aws_ebs_test.go (hunk 5 above)
Comment (ixdy) [inline pkg/volume/aws_ebs/aws_ebs_test.go]: minor nit (doesn't really need a fix either), but a better name for this probably would've been "expectedPath".

**Comment by ixdy** [comment]
LGTM, thanks!

**Pull request merged**